File: swift/Sources/CantonWalletKit/SigningDriver.swift
// Copyright (c) 2026 Victor Sima
// SPDX-License-Identifier: Apache-2.0

import CantonLedgerAPI
import CryptoKit
import Foundation

/// Produces Canton signatures over ledger-provided hashes.
///
/// Implementations decide where the private key lives: in software, in the
/// Secure Enclave, or with a custody provider. The driver never sees a full
/// transaction — only the hash the ledger asks the party to sign, which is
/// what makes hardware-backed and custody-held keys possible.
///
/// The returned signature carries no `signedBy` fingerprint; callers that
/// know the canonical fingerprint (returned by
/// `GenerateExternalPartyTopology`) fill it in.
public protocol SigningDriver: Sendable {
    /// The public key in a Ledger-API-compatible encoding.
    func publicKey() async throws -> Com_Daml_Ledger_Api_V2_SigningPublicKey

    /// Signs `bytes` (a Canton-provided hash) with the driver's private key.
    func sign(_ bytes: Data) async throws -> Com_Daml_Ledger_Api_V2_Signature
}

/// RFC 8410 SubjectPublicKeyInfo prefix for an Ed25519 public key; CryptoKit
/// only exposes the 32-byte raw form.
private let ed25519SPKIPrefix = Data([
    0x30, 0x2a, 0x30, 0x05, 0x06, 0x03, 0x2b, 0x65, 0x70, 0x03, 0x21, 0x00,
])

/// CryptoKit-backed software keys, supporting the two schemes relevant to
/// mobile wallets: Ed25519 (Canton's default) and ECDSA P-256 (the only
/// scheme the Secure Enclave signs). Both are accepted by the Ledger API;
/// verified live in the Kotlin `ExternalPartyIntegrationTest`.
public struct SoftwareSigningDriver: SigningDriver {
    public enum Algorithm: Sendable {
        case ed25519
        case ecP256
    }

    private enum Key: Sendable {
        case ed25519(Curve25519.Signing.PrivateKey)
        case ecP256(P256.Signing.PrivateKey)
    }

    private let key: Key

    public var algorithm: Algorithm {
        switch key {
        case .ed25519: .ed25519
        case .ecP256: .ecP256
        }
    }

    public static func generate(_ algorithm: Algorithm) -> SoftwareSigningDriver {
        switch algorithm {
        case .ed25519: SoftwareSigningDriver(key: .ed25519(Curve25519.Signing.PrivateKey()))
        case .ecP256: SoftwareSigningDriver(key: .ecP256(P256.Signing.PrivateKey()))
        }
    }

    public func publicKey() async throws -> Com_Daml_Ledger_Api_V2_SigningPublicKey {
        var proto = Com_Daml_Ledger_Api_V2_SigningPublicKey()
        proto.format = .derX509SubjectPublicKeyInfo
        switch key {
        case .ed25519(let key):
            proto.keyData = ed25519SPKIPrefix + key.publicKey.rawRepresentation
            proto.keySpec = .ecCurve25519
        case .ecP256(let key):
            proto.keyData = key.publicKey.derRepresentation
            proto.keySpec = .ecP256
        }
        return proto
    }

    public func sign(_ bytes: Data) async throws -> Com_Daml_Ledger_Api_V2_Signature {
        var proto = Com_Daml_Ledger_Api_V2_Signature()
        switch key {
        case .ed25519(let key):
            proto.signature = try key.signature(for: bytes)
            proto.format = .concat
            proto.signingAlgorithmSpec = .ed25519
        case .ecP256(let key):
            proto.signature = try key.signature(for: bytes).derRepresentation
            proto.format = .der
            proto.signingAlgorithmSpec = .ecDsaSha256
        }
        return proto
    }
}

#if canImport(LocalAuthentication)
/// P-256 keys generated inside the Secure Enclave: the private key is
/// non-exportable and every signature is produced by the enclave.
///
/// The Ledger API accepts P-256 external parties (verified against a live
/// participant), so this is genuine on-device self-custody. Requires real
/// hardware — unavailable in the simulator; check ``isAvailable`` first.
public struct SecureEnclaveSigningDriver: SigningDriver {
    private let key: SecureEnclave.P256.Signing.PrivateKey

    public static var isAvailable: Bool { SecureEnclave.isAvailable }

    /// Generates a fresh enclave-resident key. Pass access control flags to
    /// require biometry per signature (e.g. `.biometryCurrentSet`).
    public init(accessControl: SecAccessControlCreateFlags = [.privateKeyUsage]) throws {
        var error: Unmanaged<CFError>?
        guard
            let control = SecAccessControlCreateWithFlags(
                nil,
                kSecAttrAccessibleWhenUnlockedThisDeviceOnly,
                accessControl,
                &error
            )
        else {
            throw error!.takeRetainedValue() as Error
        }
        self.key = try SecureEnclave.P256.Signing.PrivateKey(accessControl: control)
    }

    /// Reconstructs the driver from a previously stored data representation
    /// (an opaque, enclave-bound handle — not key material).
    public init(dataRepresentation: Data) throws {
        self.key = try SecureEnclave.P256.Signing.PrivateKey(dataRepresentation: dataRepresentation)
    }

    /// Persist this alongside the party id to reuse the key across launches.
    public var dataRepresentation: Data { key.dataRepresentation }

    public func publicKey() async throws -> Com_Daml_Ledger_Api_V2_SigningPublicKey {
        var proto = Com_Daml_Ledger_Api_V2_SigningPublicKey()
        proto.format = .derX509SubjectPublicKeyInfo
        proto.keyData = key.publicKey.derRepresentation
        proto.keySpec = .ecP256
        return proto
    }

    public func sign(_ bytes: Data) async throws -> Com_Daml_Ledger_Api_V2_Signature {
        var proto = Com_Daml_Ledger_Api_V2_Signature()
        proto.signature = try key.signature(for: bytes).derRepresentation
        proto.format = .der
        proto.signingAlgorithmSpec = .ecDsaSha256
        return proto
    }
}
#endif

File: kotlin/canton-wallet-sdk/src/main/kotlin/io/github/vsima/canton/wallet/SigningDriver.kt
package io.github.vsima.canton.wallet

import com.daml.ledger.api.v2.CryptoOuterClass
import com.google.protobuf.ByteString
import java.security.KeyPair
import java.security.KeyPairGenerator
import java.security.spec.ECGenParameterSpec

/**
 * Produces Canton [CryptoOuterClass.Signature]s over ledger-provided hashes.
 *
 * Implementations decide where the private key lives: in software, in the
 * Android Keystore / StrongBox (see the android module), or in a custody
 * provider (Fireblocks, BitGo, ...). The driver never sees a full
 * transaction — only the hash the ledger asks the party to sign, which is
 * what makes hardware-backed and custody-held keys possible.
 *
 * The returned signature carries no `signed_by` fingerprint; callers that
 * know the canonical fingerprint (e.g. from
 * `GenerateExternalPartyTopologyResponse.public_key_fingerprint`) complete
 * it via [CryptoOuterClass.Signature.toBuilder].
 */
public interface SigningDriver {
    /** The public key in a Ledger-API-compatible encoding. */
    public suspend fun publicKey(): CryptoOuterClass.SigningPublicKey

    /** Signs [bytes] (a Canton-provided hash) with the driver's private key. */
    public suspend fun sign(bytes: ByteArray): CryptoOuterClass.Signature
}

/**
 * JCA-backed software keys, supporting the two schemes relevant to mobile
 * wallets: Ed25519 (Canton's default) and ECDSA P-256 (the only scheme
 * hardware enclaves sign — Apple Secure Enclave and Android StrongBox are
 * P-256-only). Both are accepted by the Ledger API; see crypto.proto.
 */
public class SoftwareSigningDriver private constructor(
    private val keyPair: KeyPair,
    public val algorithm: Algorithm,
) : SigningDriver {

    public enum class Algorithm { ED25519, EC_P256 }

    override suspend fun publicKey(): CryptoOuterClass.SigningPublicKey =
        CryptoOuterClass.SigningPublicKey.newBuilder()
            // JCA's getEncoded() is the X.509 SubjectPublicKeyInfo DER form.
            .setFormat(CryptoOuterClass.CryptoKeyFormat.CRYPTO_KEY_FORMAT_DER_X509_SUBJECT_PUBLIC_KEY_INFO)
            .setKeyData(ByteString.copyFrom(keyPair.public.encoded))
            .setKeySpec(
                when (algorithm) {
                    Algorithm.ED25519 -> CryptoOuterClass.SigningKeySpec.SIGNING_KEY_SPEC_EC_CURVE25519
                    Algorithm.EC_P256 -> CryptoOuterClass.SigningKeySpec.SIGNING_KEY_SPEC_EC_P256
                }
            )
            .build()

    override suspend fun sign(bytes: ByteArray): CryptoOuterClass.Signature {
        val (jcaName, format, spec) = when (algorithm) {
            Algorithm.ED25519 -> Triple(
                "Ed25519",
                CryptoOuterClass.SignatureFormat.SIGNATURE_FORMAT_CONCAT,
                CryptoOuterClass.SigningAlgorithmSpec.SIGNING_ALGORITHM_SPEC_ED25519,
            )
            Algorithm.EC_P256 -> Triple(
                "SHA256withECDSA",
                CryptoOuterClass.SignatureFormat.SIGNATURE_FORMAT_DER,
                CryptoOuterClass.SigningAlgorithmSpec.SIGNING_ALGORITHM_SPEC_EC_DSA_SHA_256,
            )
        }
        val signature = java.security.Signature.getInstance(jcaName).run {
            initSign(keyPair.private)
            update(bytes)
            sign()
        }
        return CryptoOuterClass.Signature.newBuilder()
            .setFormat(format)
            .setSignature(ByteString.copyFrom(signature))
            .setSigningAlgorithmSpec(spec)
            .build()
    }

    public companion object {
        public fun generate(algorithm: Algorithm): SoftwareSigningDriver {
            val keyPair = when (algorithm) {
                Algorithm.ED25519 -> KeyPairGenerator.getInstance("Ed25519").generateKeyPair()
                Algorithm.EC_P256 -> KeyPairGenerator.getInstance("EC")
                    .apply { initialize(ECGenParameterSpec("secp256r1")) }
                    .generateKeyPair()
            }
            return SoftwareSigningDriver(keyPair, algorithm)
        }
    }
}
